### ISSUE TITLE
Cannot prevent TLJH from creating new system users.

### ISSUE THREAD

**Issue by mmccarthy404** [jupyterhub/the-littlest-jupyterhub#705]
For our use case, we want JupyterHub to use PAM to authenticate users against their existing Linux user accounts. Right now, if a user named 'username' attempts to sign in, TLJH will create a new user, 'jupyter-username' rather than using the existing Linux user account 'username'. I attempted to force a change to PAM authentication (using existing accounts) with this:
```
sudo tljh-config set auth.type jupyterhub.auth.PAMAuthenticator
```
And I attempted to prevent the creation of new users with this (as documented here https://jupyterhub.readthedocs.io/en/stable/getting-started/authenticators-users-basics.html#use-localauthenticator-to-create-system-users) :
```
sudo tljh-config set c.LocalAuthenticator.create_system_users False
```
I then reloaded:
```
sudo tljh-config reload
```
But JupyterHub will still create a user account 'jupyter-username' rather than use the existing 'username' account.

**Comment by mbenguig** [comment]
Hi, sorry but I am still encountering difficulties to prevent TLJH from creating new system users. What i've done:

> sudo systemctl stop jupyterhub.service
> sudo tljh-config set auth.type jupyterhub.auth.PAMAuthenticator
> sudo tljh-config set auth.LocalAuthenticator.create_system_users False
> sudo tljh-config reload
> sudo systemctl start jupyterhub.service

.. and when i connect as an existing xxx system user... it still creates jupyter-xxx entries in /etc/passwd + /etc/group and creates a /home/jupyter-xxx folder. How to tackle this ?

**Comment by ninjit** [comment]
I'm encountering this as well, but I think the configuration change we need is somewhere else.
The "jupyter-<user>" prefix appears to be by design, as described here
https://tljh.jupyter.org/en/latest/topic/security.html

But the scenario being avoided is exactly what I (and appears @mbenguig too) want to implement:
I want TLJH to use PAM authentication to login to existing system users, without creating new ones.

Is there a configuration setting somewhere to remove the "jupyter-" prefix from all user creation/validation functions?

**Comment by mbenguig** [comment]
Any news regarding this issue ?

**Comment by fd2m** [comment]
TLDR: AFAIK it's not possible with config options. It may be possible by editing the TLJH Python code in multiple places, but I couldn't get it to work. Will probably use "normal" Jupyterhub instead.

This looked promising to me:
https://github.com/jupyterhub/the-littlest-jupyterhub/blob/0572976d25c18bc88d936f8bd8627bbfed069b53/tljh/jupyterhub_config.py#L31

This option is documented here: https://github.com/jupyterhub/systemdspawner#unit_name_template. Quote: "`{USERNAME}` and `{USERID}` in this configuration value will be expanded to the appropriate values for the user being spawned." I.e. setting the parameter to `{USERNAME}` should do the trick.

So how does this translate to a matching command for `tljh-config`? At first I thought that the structure of the `tljh-config set` command mirrors the structure of the JupyterHub configuration options. However, when the config is applied...

https://github.com/jupyterhub/the-littlest-jupyterhub/blob/0572976d25c18bc88d936f8bd8627bbfed069b53/tljh/jupyterhub_config.py#L34 
... this is achieved through a number of smaller functions which only apply a handful of parameters. So the number of config parameters useable with tljh-config implements only a (small) subset of all JupyterHub options, while also adding its own, and altering the syntax in places.

Most importantly though, `apply_config()` calls `update_user_account_config()`, which simply hardcodes it, regardless of the `config` parameter:

https://github.com/jupyterhub/the-littlest-jupyterhub/blob/0572976d25c18bc88d936f8bd8627bbfed069b53/tljh/configurer.py#L235-L236

(this is the whole function). So `tljh-config` does not allow changing this, and the only way I see to change this is by editing `jupyterhub_config.py` and `configurer.py`, which are found in the site-packages folder of the Python install running tljh, i.e. by default: `/opt/tljh/hub/lib/python3.8/site-packages/tljh`.

However I found that even this is not enough. By chance I saw the file `user_creating_spawner.py`, which contains this line:
https://github.com/jupyterhub/the-littlest-jupyterhub/blob/0572976d25c18bc88d936f8bd8627bbfed069b53/tljh/user_creating_spawner.py#L23

Even after changing this though, opening a new terminal still shows that I'm logged in as the "jupyter-{USERNAME}" user, so at this point I feel like giving up and setting up Jupyterhub from scratch.

I think I understand the safety concerns with using the local user directly when user accounts can be created on the fly. In my use case however, the users are fixed and manually administered by me, so I would disable user creation through the JupyterHub interface anyway. Having a duplicate of every local account is confusing for everyone, I think.

**Comment by mbenguig** [comment]
Pb solved following these steps (https://www.reddit.com/r/JupyterNotebooks/comments/pfah2c/can_i_setup_jupyterhub_tljh_to_use_preexisting/hb3wdeh/)

/opt/tljh/hub/lib/python3.6/site-packages/tljh/jupyterhub_config.py

Change configuration settings according to the diff below:

15c15
< c.JupyterHub.spawner_class = 'systemdspawner.SystemdSpawner'
---
> c.JupyterHub.spawner_class = UserCreatingSpawner
32,35c32
< c.SystemdSpawner.unit_name_template = 'jupyter-{USERNAME}-singleuser'
<
< c.SystemdSpawner.dynamic_users = False
< c.SystemdSpawner.username_template = '{USERNAME}'
---
> c.SystemdSpawner.unit_name_template = 'jupyter-{USERNAME}'
Next, modify the file:

/opt/tljh/hub/lib/python3.6/site-packages/tljh/configurer.py

Change configuration settings according to the diff below:

218c218
< c.SystemdSpawner.username_template = '{USERNAME}'
---
> c.SystemdSpawner.username_template = 'jupyter-{USERNAME}'
Restart Jupyterhub:

tljh-config reload

**Comment by glciampaglia** [comment]
I am encountering a similar problem when using [jub_cas_authenticator](https://github.com/cwaldbieser/jhub_cas_authenticator#just-in-time-account-creation), which provides a subclass of `LocalAuthenticator` called `CASLocalAuthenticator` for "just-in-time" account creation. 

In my case what happens is the opposite as what the OP is experiencing. I would like that, upon login, a local user called `jupyter-<id>` be created automatically (where `<id>` is the ID of the user in my institution's network --- CAS is an enterprise authentication system used a lot by universities FYI). However, what happens instead is that a local user called `<id>` is created instead, and JH return 401 Not Authorized because it cannot find `jupyter-<id>`.

Incidentally, it would be great if the username normalization (ie. adding a `jupyter-` prefix when creating a local user) could become optional. In my case it is not an issue to map JH users directly into unix users since I know in advance who are the users of the instance (I am setting up JH for a class and so I have a roster with all usernames), so it would be safe for me to turn off it entirely. (That said, of course what I am going to do is to switch to the default CASAuthenticator and create the local users manually, I was just hoping that the local authenticator would spare me from this extra manual step.)

**Comment by eingemaischt** [comment]
@consideRatio is it possible to move this to a feature request or can I create a new?
